source/code/source/playfab/PlayFabPluginManager.cpp

```diff
@@ -27,47 +29,46 @@ namespace PlayFab
 
     std::shared_ptr<IPlayFabPlugin> PlayFabPluginManager::GetPluginInternal(const PlayFabPluginContract contract, const std::string& instanceName)
     {
-        std::pair<PlayFabPluginContract, std::string> key = std::make_pair(contract, instanceName);
-        const auto pluginEntry = plugins.find(key);
-        if (pluginEntry == plugins.end())
+        PluginEntry& entry = FindOrCreatePluginEntry(contract, instanceName);
+        if (entry.plugin == nullptr)
         {
-            // Requested plugin is not in the cache, create the default one
-            std::shared_ptr<IPlayFabPlugin> pluginPtr = nullptr;
             switch (contract)
             {
             case PlayFabPluginContract::PlayFab_Serializer:
-                pluginPtr = CreatePlayFabSerializerPlugin();
+                entry.plugin = CreatePlayFabSerializerPlugin();
                 break;
             case PlayFabPluginContract::PlayFab_Transport:
-                pluginPtr = CreatePlayFabTransportPlugin();
+                entry.plugin = CreatePlayFabTransportPlugin();
                 break;
             default:
                 throw PlayFabException(PlayFabExceptionCode::PluginAmbiguity, "This contract is not supported");
                 break;
             }
-
-            plugins.insert({ key, pluginPtr });
-            return pluginPtr;
-        }
-        else
-        {
-            return pluginEntry->second;
         }
+        return entry.plugin;
     }
 
     void PlayFabPluginManager::SetPluginInternal(const std::shared_ptr<IPlayFabPlugin>& plugin, const PlayFabPluginContract contract, const std::string& instanceName)
     {
-        std::pair<PlayFabPluginContract, std::string> key = std::make_pair(contract, instanceName);
-        const auto pluginEntry = plugins.find(key);
-        if (pluginEntry == plugins.end())
-        {
-            plugins.insert({ key, std::move(plugin) });
-        }
-        else
-        {
-            plugins.erase(key);
-            plugins.insert({ key, std::move(plugin) });
-        }
+        PluginEntry& entry = FindOrCreatePluginEntry(contract, instanceName);
+        entry.plugin = plugin;
+    }
+
+    PlayFabPluginManager::PluginEntry& PlayFabPluginManager::FindOrCreatePluginEntry(PlayFabPluginContract contract, const std::string& instanceName)
+    {
+        { // LOCK pluginsMutex
+            std::unique_lock<std::mutex> lock(pluginsMutex);
+            for (PluginEntry& pluginEntry : plugins)
+            {
+                if (pluginEntry.contract == contract &&
+                    pluginEntry.name == instanceName)
+                {
+                    return pluginEntry;
+                }
+            }
+            plugins.emplace_back<PluginEntry>({contract, instanceName, nullptr});
```

Comment:
>{ [](start = 46, length = 1)

if you use emplace_back you should drop these braces (otherwise it's doing the exact same thing as push_back.

the plugins.emplace_back call also shouldn't require you to specify PluginEntry in the template parameter. I imagine that was only needed because you had the extraneous braces

Reply:
This actually doesn't seem to compile without the template syntax? Removing the template and curly brackets with just the parameters doesn't seem to map to the struct. we could make a struct outside the parentheses but that would incur similar excess copy issues.